google/cloud/firestore_v1/bulk_writer.py

```diff
@@ -0,0 +1,537 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for efficiently writing large amounts of data to the Google Cloud
+Firestore API."""
+
+import collections
+import concurrent.futures
+import enum
+import functools
+import logging
+import time
+from typing import Callable, Dict, List, NoReturn, Optional, Union
+
+from google.cloud.firestore_v1 import _helpers
+from google.cloud.firestore_v1.base_document import BaseDocumentReference
+from google.cloud.firestore_v1.base_client import BaseClient
+from google.cloud.firestore_v1.bulk_batch import BulkWriteBatch
+from google.cloud.firestore_v1.rate_limiter import RateLimiter
+from google.cloud.firestore_v1.types.firestore import BatchWriteResponse
+from google.cloud.firestore_v1.types.write import WriteResult
+
+
+logger = logging.getLogger(__name__)
+
+
+class SendMode(enum.Enum):
+    """Control flag for whether a BulkWriter should commit batches in the main
+    thread or hand that work off to an executor.
+    """
+
+    parallel = enum.auto()
+    serial = enum.auto()
+
+
+class AsyncBulkWriterMixin:
+    """
+    Mixin which contains the methods on `BulkWriter` which must only be
+    submitted to the executor (or called by functions submitted to the executor).
+    This mixin exists purely for organization and clarity of implementation
+    (e.g., there is no metaclass magic).
+
+    The entrypoint to the parallelizable code path is `_send_batch()`, which is
+    wrapped in a decorator which ensures that the `SendMode` is honored.
+    """
+
+    def _with_send_mode(fn):
+        """Decorates a method to ensure it is only called via the executor
+        (IFF the SendMode value is SendMode.parallel!).
+
+        Usage:
+
+            @_with_send_mode
+            def my_method(self):
+                parallel_stuff()
+
+            def something_else(self):
+                # Because of the decorator around `my_method`, the following
+                # method invocation:
+                self.my_method()
+                # becomes equivalent to `self._executor.submit(self.my_method)`
+                # when the send mode is `SendMode.parallel`.
+
+        Use on entrypoint methods for code paths that *must* be parallelized.
+        """
+
+        @functools.wraps(fn)
+        def wrapper(self, *args, **kwargs):
+            if self._send_mode == SendMode.parallel:
+                return self._executor.submit(lambda: fn(self, *args, **kwargs))
+            else:
+                return fn(self, *args, **kwargs)
+
+        return wrapper
```

Comment:
This decorator is only used on one method (`_send_batch()`), but in doing so simplifies multiple call locations.